### PR TITLE
Default exception handler

### PR DESCRIPTION
**Pull request by anacierdem**
~~Fix console new line bug~~

~~- Previously it was adding an extra space when the current line is exactly CONSOLE_WIDTH~~

Interrupt handler improvements

    - Remove some unnecessary nops
    - No need to disable interrupts while in the interrupt handler
    - No need to save/restore k0 & k1
    - Do not restore fcr31 register, it should be always cleared by the ex handler.
    - Store FCR31 as a 32bit value

Add cop0 & cop1 macros and start using them

Remove inline wait functions from the header

Implement and register a default exception handler

I'll be implementing a test for the interrupt handler next.

### CODE REVIEW

File: include/cop0.h
Comment (rasky) [inline include/cop0.h]: Maybe we should call them `C0_INTERRUPT_RCP` and `C0_INTERRUPT_TIMER`.

File: src/exception.c
Comment (rasky) [inline src/exception.c]: Did you try using `&abort` here? So that we don't need a specific function. But if we need to actually call `console_render()`, then a function is required.

File: src/inthandler.S
Comment (rasky) [inline src/inthandler.S]: I think it made sense to turn off the interrupt bit here. Doesn't it?

File: src/inthandler.S
Comment (rasky) [inline src/inthandler.S]: `$f26` and `$f27` I guess we need to save?

File: src/inthandler.S
Comment (rasky) [inline src/inthandler.S:233]: Uhm not sure about this. I think that, in general, we need restore `$fc31` in case the interrupt code modifies it. For instance, it contains floating point flags. I see that for the specific case of FPU exception, you need to mask some flags but maybe that should be handled specially, while still allowing the other bits to be restored.

File: src/exception.c
Comment (anacierdem) [inline src/exception.c]: We need 32 bit values here if we dont switch to a higher res.

File: src/exception.c
Comment (anacierdem) [inline src/exception.c]: Use macros here as well

File: src/exception.c
Comment (anacierdem) [inline src/exception.c]: Also combine with other addressing errors

**Review by anacierdem — COMMENTED** [review]
Self review

File: src/exception.c
Comment (anacierdem) [reply src/exception.c]: Which wont make any sense for doubles.

File: src/inthandler.S
Comment (anacierdem) [reply src/inthandler.S]: Oops..

File: src/inthandler.S
Comment (anacierdem) [reply src/inthandler.S:233]: Maybe we should set bits except the ones that needs to be reset by the handler. That would be better youre right.

File: src/exception.c
Comment (anacierdem) [reply src/exception.c]: Ill try that but I suspect it wont work

File: src/inthandler.S
Comment (anacierdem) [reply src/inthandler.S]: Why do you think so? EXL is already set at this point and I dont think an explicit disable is necessary.

File: src/exception.c
Comment (anacierdem) [reply src/exception.c]: Unfortunately it does not work. I tried to comment on why it does not work without the interrupts. It may not be 100% accurate and I think it is not worth pursuing further just to remove the extra function. We may have a look again when we update he display system someday :)

File: include/console.h
Comment (rasky) [inline include/console.h]: Self note: I want to check this on my CRT. I think borders in general help because of the typical overscan of CRTs. I'm not sure we should remove them

File: src/console.c
Comment (rasky) [inline src/console.c]: Maybe expose the margins as defines as well, so that we can reason upon them better?

File: src/exception.c
Comment (rasky) [inline src/exception.c:103]: So to clarify: this is only necessary because we need the hack of returning from the exception to call console_render, right? In principle if we fix that, we could remove these lines, right? If so, I'll add a comment for later when we get around and fix the console.

File: src/exception.c
Comment (rasky) [inline src/exception.c]: Maybe the last string can be "tmr"

File: src/interrupt.c
Comment (rasky) [inline src/interrupt.c]: INTERRUPT_2 in the command doesn't refer anymore to a macro, right? Maybe just remove it as the macro now has RCP in the name

File: src/inthandler.S
Comment (rasky) [reply src/inthandler.S]: Ok I think you're right, it's probably not required.

File: src/inthandler.S
Comment (rasky) [inline src/inthandler.S:70]: I assume you have tested this specific NOP removal on real hardware right?

File: src/inthandler.S
Comment (rasky) [inline src/inthandler.S]: OK this makes sense if we keep the current approach where the exception handler registered by the application has the task to also clear the exception (contrary to interrupts, where the interrupt is acknowledge by libdragon itself before calling the handler).

So I would maybe document at the top `exception.c` and in the comment of `register_exception_handler` that the handler must clear the exception in case it's required to avoid re-triggering.

**Review by rasky — APPROVED** [review]
Great work! I left some minor comment. 

The only thing that really concerns me is using the full resolution on a CRT for the console by default. I would be less concerned if you disabled the margin just for the exception handler, but I would expect the console by default to be more usable with some margin.

File: include/console.h
Comment (anacierdem) [reply include/console.h]: Vertical padding does not seem to be an issue on the highest res mode. We even increased the horizontal padding as it was not enough for 640x240. Please test on your screen and let's see. We are already on the limits of available screen space so I hope it looks good :)

File: src/exception.c
Comment (anacierdem) [reply src/exception.c:103]: Exactly, we can remove them from the default handler if we fix that.

File: src/interrupt.c
Comment (anacierdem) [reply src/interrupt.c]: HMm I missed this one apparently

File: src/inthandler.S
Comment (anacierdem) [reply src/inthandler.S:70]: True, it is still working on real hardware

**Comment by anacierdem** [comment]
Unfortunately current console subsytem is not easy to extend. Will need a re-write to make it more customizable. I'm not a fan of increasing the default resolution. Maybe we should do it with explicit text render calls.

File: src/inthandler.S
Comment (anacierdem) [reply src/inthandler.S]: I think it is better to leave this to the consumer indeed. Floating point exceptions are disabled by default and user may never decide to enable them.

I initially had an alternative implementation where I removed the const specifier from the exception struct so that the ex handler can change the contents to be loaded back. I started to believe that is more useful now.

If we do that, we can remove this. We will only give write access to epc and fcr31 (and maybe sr and cr) so that the custom handler won't have to use inline assembly and at the same time it means we completely inverted the control to the consumer.

What do you think @rasky?

**Comment by anacierdem** [comment]
I have also strting adding tests. Lets keep them in a single PR. It will take some time but will be worth it IMO.

File: src/inthandler.S
Comment (rasky) [reply src/inthandler.S]: I think that in general exception handlers might need to modify registers. It’s a pretty advanced use case, but why not. So maybe I’d simply drop const everywhere and document that it can be done.

**Comment by anacierdem** [comment]
Added tests & improved console paddings. The only remaining issue is the increased timer benchmark time, for which I couldnt find any lead yet.

File: src/exception.c
Comment (rasky) [inline src/exception.c]:
```suggestion
 * them directly will not work as expected as they will get overwritten.
```

File: src/exception.c
Comment (rasky) [inline src/exception.c]: I think that for *interrupts* it makes sense not to save those registers as they are callee-saved (BTW, the same I think it applies to f15-f31 but should be verified). Instead for exceptions, one might want to have a look at the whole state of all regs. So maybe the interrupt handlers should save/restore those only after it has verified that it's an exception.

Anyway, this can be improved in the future, no need to do it now.

**Review by rasky — APPROVED** [review]
Great work!

File: src/exception.c
Comment (anacierdem) [reply src/exception.c]: https://github.com/DragonMinded/libdragon/issues/129

**Comment by anacierdem** [comment]
That benchmark works fine with this state. I didn't find the cause or applied a solution though. So we better keep an eye on that. I'm mostly happy with this PR. Can merge when @rasky approves :)

**Comment by rasky** [comment]
Fine by me, I tested it everywhere and it works. I previously told you that I had an unexpected FPU exception but it turned out that was really a bug in my code so this PR actually helped me debugging a real issue.

**Pull request merged**
